### DISCUSSION TITLE
How to plot TimeSeries?

### DISCUSSION THREAD

**Discussion by wangtieqiao** [JetBrains/lets-plot-kotlin#92] [Q&A]
Is there any example notebook to show how to plot timeseries data with time in X-axis?

X-axis data could be any of the Datetime classes like  `LocalDate, LocalDateTime or ZonedDateTime` etc

**Comment by alshan** [comment]
Try using [scale_x_datetime](https://htmlpreview.github.io/?https://raw.githubusercontent.com/JetBrains/lets-plot-kotlin/master/docs/api-reference/plot-api/jetbrains.lets-plot.scale/scale_x_datetime.html).

At the moment this is the only way to plot time series. The input series can contain millis since Unix epoch. Also accepted Date, Instant and ZonedDateTime from java.time but these values will be converted to millis since Unix epoch.

LocalDate, LocalTime and LocalDateTime are not supported. 

When formatting date-time labels Lets-plot assumes UTC time-zone.

**Reply by timemanx** [reply]
> Try using [scale_x_datetime](https://htmlpreview.github.io/?https://raw.githubusercontent.com/JetBrains/lets-plot-kotlin/master/docs/api-reference/plot-api/jetbrains.lets-plot.scale/scale_x_datetime.html).
> 
> At the moment this is the only way to plot time series. The input series can contain millis since Unix epoch. Also accepted Date, Instant and ZonedDateTime from java.time but these values will be converted to millis since Unix epoch.
> 
> LocalDate, LocalTime and LocalDateTime are not supported.
> 
> When formatting date-time labels Lets-plot assumes UTC time-zone.

Is there a way now to format using the system timezone?

**Reply by alshan** [reply]
There is currently no such option, I have submitted issue #278.
As a workaround you can try manually providing breaks/labels for the axis, where the labels are formatted as required.

**Comment by wangtieqiao** [comment]
Thanks @alshan that works great! 

Another question: 
- inside of Jupyter notebook, how to make the chart interactive? (e.g. pan and zoom)
- How to add backgroud gridline? (I could not find it from any of the examples)
- And how to add title to a plot?

**Reply by alshan** [reply]
Hi, it took quite a while but we've added recently zoom/pan support via the `ggtb()` function in [Python](https://lets-plot.org/python/pages/api/lets_plot.ggtb.html) and [Kotlin](https://lets-plot.org/kotlin/api-reference/-lets--plot--kotlin/org.jetbrains.letsPlot.interact/ggtb.html).

**Comment by alshan** [comment]
Glad it worked for you. Regarding the other questions: 

To add a title use the [ggtitle](https://htmlpreview.github.io/?https://raw.githubusercontent.com/JetBrains/lets-plot-kotlin/master/docs/api-reference/plot-api/jetbrains.lets-plot.label/ggtitle.html) function.

The background gridlines are not there yet but you can generate and style a set of horizontal/vertical lines using [geom_hline](https://htmlpreview.github.io/?https://raw.githubusercontent.com/JetBrains/lets-plot-kotlin/master/docs/api-reference/plot-api/jetbrains.lets-plot.geom/geom_hline/index.html) and [geom_vline](https://htmlpreview.github.io/?https://raw.githubusercontent.com/JetBrains/lets-plot-kotlin/master/docs/api-reference/plot-api/jetbrains.lets-plot.geom/geom_vline/index.html) functions.

We don't have the zoom/pan kind of interactivity. The closest thing could be using the xlim or ylim parameters in coord_xxx function ([coord_cartesian](https://htmlpreview.github.io/?https://raw.githubusercontent.com/JetBrains/lets-plot-kotlin/master/docs/api-reference/plot-api/jetbrains.lets-plot/coord_cartesian.html) for example). Perhaps, if ipywidgets are available, a slider can be used to set these values but nobody ever tried this :).

**Comment by wangtieqiao** [comment]
Thanks! great work team! 
Any plan add those features in?

**Comment by alshan** [comment]
Yes for gridlines, panning/zooming not in the nearest future)

**Comment by wangtieqiao** [comment]
Thanks again!
a bit sad to know that zoom/pan will not be available soon.... I am a quant researcher / data scientist dealing with timesries everyday, panning/zooming is essential for my work. I see Kotlin starting to take off in analytic area, but could not find any charting library can replace what I used to in Python. Have tried BeakerX, but their wire to Kotlin kernel is broken. Anyway great work team, can not wait to see more features get into the library.

**Comment by alshan** [comment]
Thanks for the feedback! You can try the [plotly.kt](https://github.com/mipt-npm/plotly.kt) - not quite sure but it might have panning/zooming.

**Comment by alshan** [comment]
> Is there any example notebook to show how to plot timeseries data with time in X-axis?
> 
> X-axis data could be any of the Datetime classes like `LocalDate, LocalDateTime or ZonedDateTime` etc

There is a new demo in 1.3.0 with time series on X-axis: https://nbviewer.jupyter.org/github/JetBrains/lets-plot-kotlin/blob/master/docs/examples/jupyter-notebooks/formatting_axes_etc.ipynb

**Comment by loehnertz** [comment]
@wangtieqiao 
I made myself this method that works pretty well for the grid lines:
``` kotlin
    fun hlines(range: IntProgression): Feature {
        val horizontalLine: (Int) -> Feature = { yIntercept: Int ->
            geom_hline(
                yintercept = yIntercept,
                alpha = 0.1,
                color = GENERAL_MUTED_COLOR,
                size = 0.02,
            )
        }

        var horizontalLines: Feature = horizontalLine(range.first)
        for (yIntercept: Int in range.first + range.step until range.last step range.step) {
            horizontalLines += horizontalLine(yIntercept)
        }

        return horizontalLines
    }
```
It can be used like this:
```kotlin
ggplot(dataFrame) + hlines(-110 until 110 step 5)
```

Cheers 😊

**Reply by alshan** [reply]
@wangtieqiao , @loehnertz we've added gridlines in [v3.1.0](https://github.com/JetBrains/lets-plot-kotlin#what-is-new-in-310). Will be now trying to add more interactivity.

**Comment by alshan** [comment]
Support for LocalDate, LocalDateTime, ZonedDateTime etc was added in LPK v4.11.0, check this out: 
https://github.com/JetBrains/lets-plot-kotlin/blob/master/README.md#time-series-plotting
